internal/driver/glfw/loop.go

```diff
@@ -97,93 +98,137 @@ func (d *gLDriver) runGL() {
 	settingsChange := make(chan fyne.Settings)
 	fyne.CurrentApp().Settings().AddChangeListener(settingsChange)
 
-	eventTick := time.NewTicker(time.Second / 60)
 	for {
-		select {
-		case <-d.done:
-			eventTick.Stop()
-			d.Terminate()
-			l := fyne.CurrentApp().Lifecycle().(*app.Lifecycle)
-			if f := l.OnStopped(); f != nil {
-				l.QueueEvent(f)
-			}
-			return
-		case f := <-funcQueue:
-			f.f()
-			f.done <- struct{}{}
-		case <-eventTick.C:
-			d.pollEvents()
-			windowsToRemove := 0
-			for _, win := range d.windowList() {
-				w := win.(*window)
-				if w.viewport == nil {
-					continue
+		d.waitEvents()
+
+		if d.animation.HasAnimations() {
+			// Switch to running at 60 fps with d.pollEvents
+			// until we have no more animations to tick
+			t := time.NewTicker(time.Second / 60)
+			for range t.C {
+				d.pollEvents()
+				exit, animationsDone := d.runSingleFrame(settingsChange)
+				if exit {
+					return
 				}
-
-				if w.viewport.ShouldClose() {
-					windowsToRemove++
-					continue
+				if animationsDone {
+					t.Stop()
+					break
 				}
+			}
+		} else {
+			// idle mode: run single frame and sleep on d.waitEvents() above
+			exit, _ := d.runSingleFrame(settingsChange)
+			if exit {
+				return
+			}
+		}
+	}
+}
 
-				expand := w.shouldExpand
-				fullScreen := w.fullScreen
-
-				if expand && !fullScreen {
-					w.fitContent()
-					shouldExpand := w.shouldExpand
-					w.shouldExpand = false
-					view := w.viewport
+func (d *gLDriver) runSingleFrame(settingsChange <-chan fyne.Settings) (exit, animationsDone bool) {
+	// check if we're shutting down
+	select {
+	case <-d.done:
+		d.Terminate()
+		l := fyne.CurrentApp().Lifecycle().(*app.Lifecycle)
+		if f := l.OnStopped(); f != nil {
+			l.QueueEvent(f)
+		}
+		return true, false
+	default:
+		break
+	}
 
-					if shouldExpand && runtime.GOOS != "js" {
-						view.SetSize(w.shouldWidth, w.shouldHeight)
-					}
-				}
+	// run func queued to main
+	select {
+	case f := <-funcQueue:
+		f.f()
+		f.done <- struct{}{}
+	default:
+		break
+	}
 
-				d.animation.TickAnimations()
-				d.drawSingleFrame()
+	// apply settings change if any
+	select {
+	case set := <-settingsChange:
+		painter.ClearFontCache()
+		cache.ResetThemeCaches()
+		app.ApplySettingsWithCallback(set, fyne.CurrentApp(), func(w fyne.Window) {
+			c, ok := w.Canvas().(*glCanvas)
+			if !ok {
+				return
 			}
-			if windowsToRemove > 0 {
-				oldWindows := d.windowList()
-				newWindows := make([]fyne.Window, 0, len(oldWindows)-windowsToRemove)
-
-				for _, win := range oldWindows {
-					w := win.(*window)
-					if w.viewport == nil {
-						continue
-					}
-
-					if w.viewport.ShouldClose() {
-						w.visible = false
-						v := w.viewport
-
-						// remove window from window list
-						v.Destroy()
-						w.destroy(d)
-						continue
-					}
-
-					newWindows = append(newWindows, win)
-				}
+			c.applyThemeOutOfTreeObjects()
+			c.reloadScale()
+		})
+	default:
+		break
+	}
```

Comment:
I'd need to find a source for it but I am quite sure that having a single select, if possible, would be a lot more efficient with how Go deals with non-blocking channels.

Reply:
Yeah, I thought about that as well. Do we even need channels with the new threading model?

Reply:
I kept the channels for the func queue and settings listeners for now, to avoid a bigger change. (Implementing our own non-channeled queue would take a bit of work). But could be nice to move to eventually. Also I changed it so it will run all available queued funcs in one frame, vs only picking at most one per frame.

Reply:
Nice. Sounds good